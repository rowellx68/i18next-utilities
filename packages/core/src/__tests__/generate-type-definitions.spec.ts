import { describe, it, expect } from 'vitest';
import type { ResourceLanguage } from 'i18next';
import {
  generateResourceTypeDefinition,
  generateModuleTypeDefinition,
} from '../generate-type-definitions';
import type { I18NextTypedDtsOptions } from '../types';

const resources: ResourceLanguage = {
  translation: {
    'app.name': 'My App',
    'ordinal.rank_ordinal_zero': '{{count}}th',
    'ordinal.rank_ordinal_one': '{{count}}st',
    'ordinal.rank_ordinal_two': '{{count}}nd',
    'ordinal.rank_ordinal_few': '{{count}}rd',
    'ordinal.rank_ordinal_many': '{{count}}th',
    'ordinal.rank_ordinal_other': '{{count}}th',
    'plural.items_zero': '{{count}} items',
    'plural.items_one': '{{count}} item',
    'plural.items_two': '{{count}} items',
    'plural.items_few': '{{count}} items',
    'plural.items_many': '{{count}} items',
    'plural.items_other': '{{count}} items',
    'books.pagination_interval': '(0) books;(2-Inf);',
    'array.0': 'lorem',
    'array.1': 'ipsum',
    'array.2': 'dolor',
  },
};

describe('generateResourceTypeDefinition', () => {
  it('should generate the correct type definition', () => {
    const result = generateResourceTypeDefinition(resources, {
      defaultNamespace: 'translation',
    });

    expect(result).toMatchSnapshot();
  });

  it.each<I18NextTypedDtsOptions['expand']>([true, false])(
    `should generate the correct type definition when expand is %o`,
    (expand) => {
      const result = generateResourceTypeDefinition(resources, {
        defaultNamespace: 'translation',
        dts: { expand },
      });

      expect(result).toMatchSnapshot();
    }
  );
});

describe('generateModuleTypeDefinition', () => {
  it('should generate the correct type definition', () => {
    const result = generateModuleTypeDefinition(
      'virtual:i18next-typed-loader',
      ['en', 'es', 'en-GB']
    );

    expect(result).toMatchSnapshot();
  });
});
