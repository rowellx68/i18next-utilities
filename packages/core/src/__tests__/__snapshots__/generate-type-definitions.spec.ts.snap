// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generateModuleTypeDefinition > should generate the correct type definition 1`] = `
"import { CustomTypeOptions } from 'i18next'

declare module 'virtual:i18next-typed-loader' {
  const resources: {
    'en': CustomTypeOptions['resources']
    'es': CustomTypeOptions['resources']
    'en-GB': CustomTypeOptions['resources']
  }

  export default resources
}
"
`;

exports[`generateResourceTypeDefinition > should generate the correct type definition 1`] = `
"import 'i18next'

type GeneratedResources = {
  'translation': {
    'app.name': string
    'array': string
    'array.0': string
    'array.1': string
    'array.2': string
    'books.pagination': string
    'books.pagination_interval': string
    'ordinal.rank': string
    'ordinal.rank_ordinal_few': string
    'ordinal.rank_ordinal_many': string
    'ordinal.rank_ordinal_one': string
    'ordinal.rank_ordinal_other': string
    'ordinal.rank_ordinal_two': string
    'ordinal.rank_ordinal_zero': string
    'plural.items': string
    'plural.items_few': string
    'plural.items_many': string
    'plural.items_one': string
    'plural.items_other': string
    'plural.items_two': string
    'plural.items_zero': string
  }
}

type FlatGeneratedResources<TResource, NS extends keyof TResource> = {
  [Property in keyof TResource[NS] as \`\${NS & string}:\${Property & string}\`]: TResource[NS][Property]
}

declare module 'i18next' {
  interface CustomTypeOptions {
    defaultNS: 'translation'
    resources: GeneratedResources
      & { 'translation': FlatGeneratedResources<GeneratedResources, 'translation'> }
  }
}
"
`;

exports[`generateResourceTypeDefinition > should generate the correct type definition when expand is false 1`] = `
"import 'i18next'

type GeneratedResources = {
  'translation': {
    'app.name': string
    'array': string
    'books.pagination': string
    'ordinal.rank': string
    'plural.items': string
  }
}

type FlatGeneratedResources<TResource, NS extends keyof TResource> = {
  [Property in keyof TResource[NS] as \`\${NS & string}:\${Property & string}\`]: TResource[NS][Property]
}

declare module 'i18next' {
  interface CustomTypeOptions {
    defaultNS: 'translation'
    resources: GeneratedResources
      & { 'translation': FlatGeneratedResources<GeneratedResources, 'translation'> }
  }
}
"
`;

exports[`generateResourceTypeDefinition > should generate the correct type definition when expand is true 1`] = `
"import 'i18next'

type GeneratedResources = {
  'translation': {
    'app.name': string
    'array': string
    'array.0': string
    'array.1': string
    'array.2': string
    'books.pagination': string
    'books.pagination_interval': string
    'ordinal.rank': string
    'ordinal.rank_ordinal_few': string
    'ordinal.rank_ordinal_many': string
    'ordinal.rank_ordinal_one': string
    'ordinal.rank_ordinal_other': string
    'ordinal.rank_ordinal_two': string
    'ordinal.rank_ordinal_zero': string
    'plural.items': string
    'plural.items_few': string
    'plural.items_many': string
    'plural.items_one': string
    'plural.items_other': string
    'plural.items_two': string
    'plural.items_zero': string
  }
}

type FlatGeneratedResources<TResource, NS extends keyof TResource> = {
  [Property in keyof TResource[NS] as \`\${NS & string}:\${Property & string}\`]: TResource[NS][Property]
}

declare module 'i18next' {
  interface CustomTypeOptions {
    defaultNS: 'translation'
    resources: GeneratedResources
      & { 'translation': FlatGeneratedResources<GeneratedResources, 'translation'> }
  }
}
"
`;
